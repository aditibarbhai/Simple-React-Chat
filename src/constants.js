module.exports = {
    PORT: 8080,
    connection: 'connection',
    disconnect: 'disconnect',
    CREATE_USER: 'CREATE_USER',
    SEND_MESSAGE: 'SEND_MESSAGE',
    RECEIVE_MESSAGE: 'RECEIVE_MESSAGE',
    WHISPER: 'WHISPER',
    USERNAMES: 'USERNAMES'
};