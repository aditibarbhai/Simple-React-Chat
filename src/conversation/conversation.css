.header {
    width: 80% !important;
}

.header .face-icon {
    margin-right: 15px;
}

.send-button svg {
    margin-left: 8px;
}

.conversation-form-wrapper {
    background-color: #ffffff;
    bottom: 0px;
    position: fixed;
    min-height: 60px;
    width: 80%;
    -webkit-box-shadow: 0px -5px 10px 0px rgba(0,0,0,0.2);
    -moz-box-shadow: 0px -5px 10px 0px rgba(0,0,0,0.2);
    box-shadow: 0px -5px 10px 0px rgba(0,0,0,0.2);
}

.private-conversation-form {
    display: flex;
    justify-content: space-evenly;
    padding: 10px 0;
    width: 100%;
}

.messages {
    margin-top: 65px;
    max-height: 80vh;
    overflow: scroll;
    width: 100%;
}

.friend-message {
    background-color: #E8EAF6;
}

.friend-message strong {
    color: #7986CB;
}

.message-input {
    width: 85%;
    margin-left: 10px;
}

.send-message {
    width: 11%;
}