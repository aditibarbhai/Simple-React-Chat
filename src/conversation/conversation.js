import React, { Component } from 'react';
import { connect } from 'react-redux';
import constants from '../constants';
import SendIcon from '@material-ui/icons/Send';
import { 
    FormControl,
    FormHelperText,
    Input,
    Button,
    List,
    ListItem,
    ListItemText,
    AppBar,
    Toolbar,
    Typography
} from '@material-ui/core';
import Face from '@material-ui/icons/Face';

import { sendMessage, updateMessages, updateError } from '../actions/chat-actions';
import './conversation.css';

export class Conversation extends Component {
    constructor(props) {
        super(props);

        this.state = {
            message: '',
            newMessage: null
        };

        this.messagesRef = null;

        this.handleMessageChange = this.handleMessageChange.bind(this);
        this.submitMessage = this.submitMessage.bind(this);
        this.scrollToBottomOfMessags = this.scrollToBottomOfMessages.bind(this);
        this.sendMessageCallback = this.sendMessageCallback.bind(this);
    }

    componentWillReceiveProps(nextProps) {
        this.scrollToBottomOfMessages();
    }

    scrollToBottomOfMessages() {
        this.messagesRef.scrollTop = this.messagesRef.scrollHeight;
    }

    handleMessageChange(event) {
        this.setState({
            message: event.target.value
        });
    }

    sendMessageCallback(data) {
        if (data.messageSent) {
            this.props.updateMessages(this.state.newMessage);
        } else {
            this.props.updateError({
                visible: true,
                message: data.error
            });
        }
    }

    submitMessage(event) {
        event.preventDefault();
        
        if (this.state.message) {
            const newMessage = {
                message: this.state.message,
                to: this.props.selectedUser,
                from: this.props.myUsername
            };

            this.props.socket.emit(constants.SEND_MESSAGE, newMessage, this.sendMessageCallback);

            this.setState({
                message: '',
                newMessage
            });
        }
    }

    renderMessages(messages) {
        return messages
            .filter(message => message.from === this.props.selectedUser || message.to ===  this.props.selectedUser)
            .map((data, index) => {
                let styles = '';

                if (data.from !== this.props.myUsername) {
                    styles += ' friend-message ';
                }

                return (
                    <ListItem className={styles} key={index + data.to}>
                        <ListItemText>
                            <Typography variant="body1" gutterBottom align="left">
                                <strong> { data.from }: </strong> { data.message }
                            </Typography>
                        </ListItemText>
                    </ListItem>
                );
            });
    }

    render() {
        return (
            <div className="private-conversation">
                <AppBar className="header"> 
                    <Toolbar>
                        <Face className="face-icon" />
                        <Typography variant="title" color="inherit">
                        { this.props.selectedUser }
                        </Typography>
                    </Toolbar>
                </AppBar>
                <div className="messages" ref={(el) => { this.messagesRef = el; }}>
                    <List dense={true}>
                        { this.renderMessages(this.props.messages) }
                    </List>
                </div>
                <div className="conversation-form-wrapper">
                    <form className="private-conversation-form">
                        <FormControl className="message-input">
                            <Input 
                                type="text" 
                                value={this.state.message}
                                onChange={this.handleMessageChange}
                            />
                            {
                                this.props.error.visible &&
                                <FormHelperText error={this.props.error.visible} > 
                                    { this.props.error.message } 
                                </FormHelperText>
                            }
                        </FormControl>
                        <FormControl className="send-message">
                            <Button 
                                className="send-button" 
                                color="primary" 
                                onClick={this.submitMessage}
                                type="submit"
                                variant="contained" 
                            >
                                Send
                                <SendIcon />
                            </Button>
                        </FormControl>
                    </form>
                </div>
            </div>
        );
    }
}

Conversation.propTypes = {};

const mapStateToProps = state => ({
    selectedUser: state.selectedUser,
    myUsername: state.myUsername,
    socket: state.socket,
    error: state.error,
    messages: state.messages
});

const mapDispatchToProps = dispatch => ({
    sendMessage: message => dispatch(sendMessage(message)),
    updateMessages: message => dispatch(updateMessages(message)),
    updateError: error => dispatch(updateError(error))
});

export default connect(mapStateToProps, mapDispatchToProps)(Conversation);