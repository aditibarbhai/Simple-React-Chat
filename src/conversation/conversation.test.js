import React from 'react';
import { Conversation } from './conversation';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import constants from '../constants';

describe('Conversation', () => {
    let props = {};

    beforeEach(() => {
        props = {
            selectedUser: 'testUser',
            myUsername: 'myTestUsername',
            error: {
                visible: false,
                message: ''
            },
            socket: {
                emit: jest.fn()
            },
            messages: [],
            updateMessages: jest.fn(),
            updateError: jest.fn()
        };
    });

    it ('should render without errors', () => {
        const tree = renderer
            .create(<Conversation {...props} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it ('submitMessage should send the message to selected user', () => {
        const wrapper = shallow(<Conversation {...props} />);

        wrapper.setState({ message: 'test message' });
        wrapper.instance().submitMessage({ preventDefault: jest.fn() });

        expect(props.socket.emit).toHaveBeenCalledWith(constants.SEND_MESSAGE, {
            message: 'test message',
            to: props.selectedUser,
            from: props.myUsername
        }, expect.any(Function));
        expect(wrapper.state().message).toBe('');
        expect(wrapper.state().newMessage).toBeDefined();
    });

    it ('sendMessageCallback should update messages if message sent', () => {
        const wrapper = shallow(<Conversation {...props} />);

        wrapper.instance().sendMessageCallback({ messageSent: true });

        expect(props.updateMessages).toHaveBeenCalledWith(expect.any(Object));
    });

    it ('sendMessageCallback should update error if message not sent', () => {
        const wrapper = shallow(<Conversation {...props} />);

        wrapper.instance().sendMessageCallback({ error: 'User disconnected' });

        expect(props.updateError).toHaveBeenCalledWith({ visible: true, message: 'User disconnected' });
    });

    it ('handleMessageChange should update state.message', () => {
        const wrapper = shallow(<Conversation {...props} />);

        expect(wrapper.state().message).toBe('');
        
        wrapper.instance().handleMessageChange({ target: { value: 'hello!' } });
        
        expect(wrapper.state().message).toBe('hello!');
    });

    it ('sshould update scroll position as new messages are added', () => {
        const wrapper = mount(<Conversation {...props} />),
            scrollSpy = jest.spyOn(Conversation.prototype, 'scrollToBottomOfMessages');
        
        wrapper.setProps({
            messages: [{
                message: 'test message',
                to: 'testUser',
                from: 'myUsername'
            }]
        });

        expect(scrollSpy).toHaveBeenCalled();
    });
});