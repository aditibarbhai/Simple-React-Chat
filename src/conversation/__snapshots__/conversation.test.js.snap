// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conversation should render without errors 1`] = `
<div
  className="private-conversation"
>
  <header
    className="MuiPaper-root-9 MuiPaper-elevation4-15 MuiAppBar-root-1 MuiAppBar-positionFixed-2 MuiAppBar-colorPrimary-7 mui-fixed header"
  >
    <div
      className="MuiToolbar-root-36 MuiToolbar-gutters-37"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-38 face-icon"
        color={undefined}
        focusable="false"
        viewBox="0 0 24 24"
      >
        <g>
          <path
            d="M9 11.75c-.69 0-1.25.56-1.25 1.25s.56 1.25 1.25 1.25 1.25-.56 1.25-1.25-.56-1.25-1.25-1.25zm6 0c-.69 0-1.25.56-1.25 1.25s.56 1.25 1.25 1.25 1.25-.56 1.25-1.25-.56-1.25-1.25-1.25zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8 0-.29.02-.58.05-.86 2.36-1.05 4.23-2.98 5.21-5.37C11.07 8.33 14.05 10 17.42 10c.78 0 1.53-.09 2.25-.26.21.71.33 1.47.33 2.26 0 4.41-3.59 8-8 8z"
          />
        </g>
      </svg>
      <h2
        className="MuiTypography-root-44 MuiTypography-title-50 MuiTypography-colorInherit-63"
      >
        testUser
      </h2>
    </div>
  </header>
  <div
    className="messages"
  >
    <ul
      className="MuiList-root-68 MuiList-dense-70 MuiList-padding-69"
    />
  </div>
  <div
    className="conversation-form-wrapper"
  >
    <form
      className="private-conversation-form"
    >
      <div
        className="MuiFormControl-root-72 message-input"
      >
        <div
          className="MuiInput-root-76 MuiInput-formControl-77 MuiInput-underline-80"
        >
          <input
            aria-invalid={false}
            aria-required={false}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-84"
            defaultValue={undefined}
            disabled={false}
            id={undefined}
            name={undefined}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root-72 send-message"
      >
        <button
          className="MuiButtonBase-root-111 MuiButton-root-89 MuiButton-contained-98 MuiButton-containedPrimary-99 MuiButton-raised-101 MuiButton-raisedPrimary-102 send-button"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="submit"
        >
          <span
            className="MuiButton-label-90"
          >
            Send
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-38"
              color={undefined}
              focusable="false"
              viewBox="0 0 24 24"
            >
              <g>
                <path
                  d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
                />
              </g>
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-114"
          />
        </button>
      </div>
    </form>
  </div>
</div>
`;
